grafter.repository/src/clj/grafter_2/rdf4j/repository.clj

```diff
@@ -137,20 +140,127 @@
+                        user-agent]
+    :or {max-conn-total 10
+         max-conn-per-route 10
+         conn-time-to-live 60
+         conn-time-to-live-timeunit TimeUnit/SECONDS
+         user-agent "grafter"
+         }}]
+  (-> (HttpClients/custom)
+      ;;(.useSystemProperties) ;; we could do this but better to be explict in clj
+
+      ;; NOTE there are additional
+      (.setMaxConnTotal max-conn-total)
+      (.setMaxConnPerRoute max-conn-per-route)
+      (.setUserAgent user-agent)
+      (.setConnectionTimeToLive conn-time-to-live conn-time-to-live-timeunit)))
+
+(defn make-default-thread-pool
+  "Create a new thread-pool suitable for use with RDF4j's
+  SPARQLRepository.
+
+  Takes a map of key value pairs:
+
+  - `:grafter.http/io-thread-pool-size` (default 10) the fixed size of the thread pool to read background IO
+
+  "
+  [{:grafter.http/keys [io-thread-pool-size] :or {io-thread-pool-size 10}}]
+  ;; NOTE the code here is essentially a port of the code found in
+  ;; RDF4j:
+  ;; https://github.com/eclipse/rdf4j/blob/0b74c317c4e7508ca4518eb9486dbc292d299d41/core/http/client/src/main/java/org/eclipse/rdf4j/http/client/SharedHttpClientSessionManager.java#L167
+  (let [backing-thread-factory (Executors/defaultThreadFactory)]
+    (Executors/newScheduledThreadPool ;; NOTE this operates as a fixed size pool!
+     io-thread-pool-size
+     (reify java.util.concurrent.ThreadFactory
+       (newThread [this runnable]
+         (let [thread (.newThread backing-thread-factory runnable)]
+           (doto thread
+             (.setName (str "grafter-http-thread-" (.getName thread))))))))))
+
+(def ^:private default-thread-pool
+  (memoize make-default-thread-pool))
+
+(defn make-shared-session-manager
+  "A default shared session manager, configuring the behaviour of the
+  http connection and thread pooling used inside the HttpClient and
+  RDF4j.
+
+  Accepts a map of key value pairs:
+
+  - `:grafter/http-client-builder` (optional) if supplied overrides the default shared HttpClientBuilder (see `make-http-client-builder`)
+  - `:grafter/thread-pool` (optional) if supplied overrides the default ScheduledThreadPoolExecutor.
+  "
+
+  [{:grafter/keys [http-client-builder thread-pool] :as opts}]
+  (let [thread-pool (or thread-pool (make-default-thread-pool opts))
+        http-client (.build (or http-client-builder (make-http-client-builder opts)))]
+    (reify org.eclipse.rdf4j.http.client.HttpClientSessionManager
+      (getHttpClient [t]
+        http-client)
+      (createSPARQLProtocolSession [t query-url update-url]
+        (grafter_2.rdf.SPARQLSession. query-url update-url http-client thread-pool))
+      (createRDF4JProtocolSession [t server-url]
+        (assert false "createRDF4JProtocolSession not implemented on grafter HttpClientSessionManager"))
+
+      (shutDown [t]
+        (.shutdown thread-pool)
+        (.awaitTermination thread-pool 10 TimeUnit/SECONDS)
+
+        ;; TODO should also probably close sessions
```

Comment:
Does this still need to happen?

Reply:
Ahh yes thanks for spotting! 🙇  

Strictly speaking it's not a big deal, as we will rarely shut these down, but I did mean to do this, so I've pushed some code which closes these on shutdown along with some basic tests.